projects/dataset_recorder/robot_recorder.h

```diff
@@ -0,0 +1,225 @@
+            roll = angles[2];
+        } catch (std::exception &e) {
+            LOGE << "Could not read from IMU: " << e.what();
+        }
+
+        // read speed and steering angle from robot
+        float botSpeed = NAN;
+        degree_t turnAngle{ NAN };
+        try {
+            std::tie(botSpeed, turnAngle) = robot.readRemoteControl();
+        } catch (std::exception &e) {
+            // if we can't read speed or angle, we just write nan values
+            LOGE << "Could not read speed and steering angle from robot : " << e.what();
+        }
+
+        // Poll for GPS data
+        if (const auto optData = m_GPS.read()) {
+            const auto &gpsData = optData.value();
+            const auto &coord = gpsData.coordinate;
+            int gpsQual = (int) gpsData.gpsQuality; // gps quality
+
+            // output results
+            LOGD << std::setprecision(10)
+                << "GPS quality: " << gpsQual
+                << " latitude: " << coord.lat.value()
+                << " longitude: " << coord.lon.value()
+                << " num sats: " << gpsData.numberOfSatellites
+                << " time: " << time.value() << std::endl;
+
+            // converting to UTM
+            const auto utm = MapCoordinate::latLonToUTM(coord);
+            m_Recorder.record(utm.toVector(), degree_t{ yaw }, m_Frame,
+                              pitch.value(), roll.value(), botSpeed,
+                              turnAngle.value(), utm.zone,
+                              (int) gpsData.gpsQuality,
+                              millimeter_t{ gpsData.horizontalDilution }.value(),
+                              time.value());
+        } else {
+            /*
+             * No new data was available. Indicate this by writing NaNs to the
+             * CSV file. We can always estimate these missing values post hoc
+             * with interpolation.
+             */
+            m_Recorder.record(Vector3<millimeter_t>::nan(), degree_t{ yaw },
+                              m_Frame, pitch.value(), roll.value(), botSpeed,
+                              turnAngle.value(), "", -1, NAN, time.value());
+        }
+
+        return time;
+    }
+
+    const auto &getGPSData() const
+    {
+        return m_GPSData;
+    }
+
+private:
+    GPS::GPSReader &m_GPS;
+    GPS::GPSData m_GPSData;
+    std::unique_ptr<Video::Input> m_Camera;
+    cv::Mat m_Frame;
+    Navigation::ImageDatabase m_Database;
+    Navigation::ImageDatabase::RouteRecorder<Navigation::ImageDatabase::VideoFileWriter> m_Recorder;
+    Stopwatch m_Stopwatch;
+
+    static std::unique_ptr<Video::Input> getCamera()
+    {
+#ifdef DUMMY_CAMERA
+        LOGW << "USING DUMMY CAMERA!!!!!!!!";
+        return std::make_unique<Video::RandomInput<>>(cv::Size{ 360, 100 });
+#else
+        return Video::getPanoramicCamera();
+#endif
+    }
+
+    units::frequency::hertz_t getFrameRate() const
+    {
+#ifdef DUMMY_CAMERA
+        return units::frequency::hertz_t{ 30 };
```

Comment:
Again, it would be much cleaner to just add this to ``Video::RandomInput`` (or make a derived version just used here or something) what with this already being a virtual method

Reply:
True! Done